File: engines/r-scripts/efa.R
# efa.R - Exploratory Factor Analysis for Method Studio
#
# Expected environment variables (set by wrapper.R from request data):
#   variables        : list/character vector of column names to analyze (min 3)
#   <col_name>       : each column's data injected as a variable (list from JSON)
#   nFactors         : integer, number of factors to extract (0 = auto via Kaiser)
#   rotation         : character, rotation method: "varimax" | "promax" | "oblimin" |
#                      "quartimax" | "none" (default: "varimax")
#   extractionMethod : character, extraction method: "minres" | "ml" | "pa" | "wls" | "gls"
#                      (default: "minres")
#   kmoTest          : logical, compute KMO measure (default: TRUE)
#   bartlettTest     : logical, compute Bartlett's test (default: TRUE)
#   screePlot        : logical, draw scree plot (default: TRUE)
#   factorScores     : logical, compute factor scores (default: FALSE)
#   suppressLoadings : numeric, suppress loadings below this value (default: 0.3)
#   missingValues    : character, "exclude-listwise" | "pairwise" (default: "exclude-listwise")
#   alpha            : numeric, significance level for Bartlett's test (default: 0.05)
#
# Returns a named list assigned to `result` with:
#   adequacy           : list with KMO and Bartlett's test results
#   eigenvalues        : numeric vector
#   variance_explained : list of per-factor variance info
#   loadings           : list of per-variable loading rows
#   communalities      : named list (h2 per variable)
#   uniqueness         : named list (u2 per variable)
#   n_factors          : integer, number of factors extracted
#   rotation           : character, rotation used
#   fm                 : character, extraction method used
#   n                  : integer, number of observations
#   n_vars             : integer, number of variables analysed
#   variable_names     : character vector of analysed variable names
#   interpretation     : character, summary interpretation
#
# Dependencies: psych (loaded by caller via packages field)

# ---- Input validation ----

if (!exists("variables") || length(variables) == 0) {
  stop("Variable 'variables' is required - specify column names to analyze")
}

if (!requireNamespace("psych", quietly = TRUE)) {
  stop("Package 'psych' is required for EFA. Install it with install.packages('psych').")
}

# Variables come as a list from JSON - extract as character vector
var_names <- if (is.list(variables)) unlist(variables) else variables

# Build data frame from injected column variables (same pattern as descriptives.R)
df_list <- list()
for (vn in var_names) {
  if (!exists(vn)) {
    stop(paste0("Column '", vn, "' not found in data"))
  }
  x_raw <- get(vn)
  df_list[[vn]] <- as.numeric(if (is.list(x_raw)) unlist(x_raw) else x_raw)
}
df_raw <- as.data.frame(df_list, stringsAsFactors = FALSE)

# Enforce minimum variables
if (ncol(df_raw) < 3) {
  stop(paste0("EFA requires at least 3 numeric variables; only ", ncol(df_raw), " provided."))
}

# ---- Options ----

# nFactors: 0 or missing means auto-detect
if (!exists("nFactors") || is.null(nFactors)) {
  nfactors_req <- 0L
} else {
  # JSON numbers come as lists; extract scalar
  nfactors_raw <- if (is.list(nFactors)) nFactors[[1]] else nFactors
  nfactors_req <- as.integer(nfactors_raw)
}

rotation_opt <- "varimax"
if (exists("rotation") && !is.null(rotation)) {
  rotation_opt <- if (is.list(rotation)) rotation[[1]] else rotation
}

fm_opt <- "minres"
if (exists("extractionMethod") && !is.null(extractionMethod)) {
  fm_raw <- if (is.list(extractionMethod)) extractionMethod[[1]] else extractionMethod
  # Map spec values to psych values
  fm_opt <- switch(fm_raw,
    "minres" = "minres",
    "ml"     = "ml",
    "pa"     = "pa",
    "wls"    = "wls",
    "gls"    = "gls",
    "minres"  # default fallback
  )
}

do_kmo       <- TRUE
if (exists("kmoTest") && !is.null(kmoTest)) {
  do_kmo <- as.logical(if (is.list(kmoTest)) kmoTest[[1]] else kmoTest)
}

do_bartlett  <- TRUE
if (exists("bartlettTest") && !is.null(bartlettTest)) {
  do_bartlett <- as.logical(if (is.list(bartlettTest)) bartlettTest[[1]] else bartlettTest)
}

do_scree     <- TRUE
if (exists("screePlot") && !is.null(screePlot)) {
  do_scree <- as.logical(if (is.list(screePlot)) screePlot[[1]] else screePlot)
}

alpha_opt <- 0.05
if (exists("alpha") && !is.null(alpha)) {
  alpha_opt <- as.numeric(if (is.list(alpha)) alpha[[1]] else alpha)
}

missing_opt <- "exclude-listwise"
if (exists("missingValues") && !is.null(missingValues)) {
  missing_opt <- if (is.list(missingValues)) missingValues[[1]] else missingValues
}

# Validate rotation
valid_rotations <- c("varimax", "promax", "oblimin", "none", "quartimax")
if (!rotation_opt %in% valid_rotations) {
  warning(paste0("Unknown rotation '", rotation_opt, "'; defaulting to 'varimax'."))
  rotation_opt <- "varimax"
}

# Validate extraction method
valid_fm <- c("minres", "ml", "pa", "wls", "gls")
if (!fm_opt %in% valid_fm) {
  warning(paste0("Unknown extraction method '", fm_opt, "'; defaulting to 'minres'."))
  fm_opt <- "minres"
}

# ---- Handle missing values ----

if (missing_opt == "pairwise") {
  # Use pairwise complete observations for correlation; keep all rows
  df <- df_raw
  use_str <- "pairwise.complete.obs"
} else {
  # Listwise deletion (default)
  df <- df_raw[complete.cases(df_raw), , drop = FALSE]
  use_str <- "everything"
}

n_obs <- nrow(df)

if (n_obs < ncol(df) + 1) {
  stop(paste0(
    "Insufficient observations (", n_obs, ") for ", ncol(df),
    " variables. Need at least ", ncol(df) + 1, " complete cases."
  ))
}

# ---- Adequacy Tests ----

cor_matrix <- cor(df, use = if (missing_opt == "pairwise") "pairwise.complete.obs" else "everything")

# KMO (Kaiser-Meyer-Olkin measure of sampling adequacy)
kmo_overall  <- NULL
kmo_per_var  <- NULL
kmo_interp   <- NULL

kmo_interpret <- function(msa) {
  if (msa >= 0.90) {
    "Marvelous"
  } else if (msa >= 0.80) {
    "Meritorious"
  } else if (msa >= 0.70) {
    "Middling"
  } else if (msa >= 0.60) {
    "Mediocre"
  } else if (msa >= 0.50) {
    "Miserable"
  } else {
    "Unacceptable"
  }
}

if (do_kmo) {
  kmo_result  <- psych::KMO(cor_matrix)
  kmo_overall <- round(unname(kmo_result$MSA), 4)
  kmo_per_var <- round(kmo_result$MSAi, 4)
  kmo_interp  <- kmo_interpret(kmo_overall)
}

# Bartlett's Test of Sphericity
bart_chi_sq  <- NULL
bart_df      <- NULL
bart_p_value <- NULL
bart_sig     <- NULL
bart_note    <- NULL

if (do_bartlett) {
  bart_result  <- psych::cortest.bartlett(cor_matrix, n = n_obs)
  bart_chi_sq  <- round(unname(bart_result$chisq), 4)
  bart_df      <- unname(bart_result$df)
  bart_p_value <- round(unname(bart_result$p.value), 8)
  bart_sig     <- bart_p_value < alpha_opt
  bart_note    <- if (bart_sig) {
    "Bartlett's test is significant: factor analysis is appropriate."
  } else {
    "Bartlett's test is NOT significant: factor analysis may not be appropriate."
  }
}

adequacy <- list(
  kmo = if (do_kmo) list(
    overall        = kmo_overall,
    per_variable   = as.list(kmo_per_var),
    interpretation = kmo_interp
  ) else NULL,
  bartlett = if (do_bartlett) list(
    chi_square  = bart_chi_sq,
    df          = bart_df,
    p_value     = bart_p_value,
    significant = bart_sig,
    note        = bart_note
  ) else NULL
)

# ---- Determine nfactors ----

if (nfactors_req <= 0L) {
  # Kaiser criterion: eigenvalues > 1
  eig_vals <- eigen(cor_matrix)$values
  nfactors <- max(1L, as.integer(sum(eig_vals > 1.0)))
} else {
  nfactors <- nfactors_req
  max_factors <- max(1L, floor((ncol(df) - 1) / 2))
  if (nfactors > max_factors) {
    warning(paste0(
      "nFactors (", nfactors, ") exceeds maximum recommended (",
      max_factors, "); setting to ", max_factors
    ))
    nfactors <- max_factors
  }
}

# ---- EFA ----

rotate_arg <- if (rotation_opt == "none") "none" else rotation_opt

# Try EFA with requested method, fall back to more robust methods if needed
fa_result <- tryCatch({
  psych::fa(
    r        = df,
    nfactors = nfactors,
    rotate   = rotate_arg,
    fm       = fm_opt,
    scores   = "none",
    warnings = FALSE
  )
}, error = function(e) {
  # If ML or other method fails, try minres
  if (fm_opt != "minres") {
    warning(paste0("Method '", fm_opt, "' failed, trying 'minres': ", e$message))
    tryCatch({
      psych::fa(r = df, nfactors = nfactors, rotate = rotate_arg, fm = "minres", scores = "none", warnings = FALSE)
    }, error = function(e2) {
      # If minres also fails, try principal axis factoring
      warning(paste0("Method 'minres' also failed, trying 'pa': ", e2$message))
      psych::fa(r = df, nfactors = nfactors, rotate = rotate_arg, fm = "pa", scores = "none", warnings = FALSE)
    })
  } else {
    # minres failed, try pa
    warning(paste0("Method 'minres' failed, trying 'pa': ", e$message))
    psych::fa(r = df, nfactors = nfactors, rotate = rotate_arg, fm = "pa", scores = "none", warnings = FALSE)
  }
})

# ---- Extract outputs ----

# Loadings (vars x factors)
fa_loadings_raw <- unclass(fa_result$loadings)
factor_names    <- paste0("F", seq_len(nfactors))
rownames(fa_loadings_raw) <- colnames(df)
colnames(fa_loadings_raw) <- factor_names

# Convert to a list of lists for JSON-friendly serialisation
loadings_list <- lapply(seq_len(nrow(fa_loadings_raw)), function(i) {
  row <- as.list(round(fa_loadings_raw[i, ], 4))
  c(list(variable = rownames(fa_loadings_raw)[i]), row)
})

# Communalities and uniqueness
communalities <- round(fa_result$communality, 4)
uniqueness    <- round(fa_result$uniquenesses, 4)
names(communalities) <- colnames(df)
names(uniqueness)    <- colnames(df)

# Eigenvalues (from correlation matrix)
eigenvalues <- round(eigen(cor_matrix)$values, 4)

# Variance explained table
ss_loadings <- colSums(fa_loadings_raw^2)
prop_var    <- ss_loadings / ncol(df)
cumul_var   <- cumsum(prop_var)

variance_explained <- lapply(seq_len(nfactors), function(i) {
  list(
    factor      = factor_names[i],
    ss_loadings = round(ss_loadings[i], 4),
    prop_var    = round(prop_var[i], 4),
    cumul_var   = round(cumul_var[i], 4)
  )
})

# ---- Scree Plot ----

if (do_scree) {
  n_eig_to_plot <- min(ncol(df), 20L)
  plot(
    x    = seq_len(n_eig_to_plot),
    y    = eigenvalues[seq_len(n_eig_to_plot)],
    type = "b",
    pch  = 20,
    col  = "#4C72B0",
    xlab = "Factor Number",
    ylab = "Eigenvalue",
    main = "Scree Plot",
    ylim = c(0, max(eigenvalues[seq_len(n_eig_to_plot)]) * 1.1),
    xaxt = "n"
  )
  axis(1, at = seq_len(n_eig_to_plot))
  abline(h = 1, col = "#DD8452", lwd = 1.5, lty = 2)
  abline(v = nfactors, col = "#55A868", lwd = 1.5, lty = 3)
  legend(
    "topright",
    legend = c("Eigenvalue = 1", paste0("Retained (", nfactors, " factors)")),
    col    = c("#DD8452", "#55A868"),
    lty    = c(2, 3),
    lwd    = 1.5,
    bty    = "n"
  )
}

# ---- Interpretation ----

adequacy_ok <- (!is.null(kmo_overall) && kmo_overall >= 0.60) &&
               (!is.null(bart_sig) && bart_sig)
var_pct     <- round(cumul_var[nfactors] * 100, 1)

kmo_str  <- if (!is.null(kmo_overall)) {
  sprintf("KMO = %.3f (%s). ", kmo_overall, kmo_interp)
} else {
  ""
}

bart_p_cmp <- if (!is.null(bart_sig) && bart_sig) "<" else ">="
bart_str <- if (!is.null(bart_sig)) {
  sprintf("Bartlett's test: chi2(%d) = %.2f, p %s %.4f. ",
    bart_df, bart_chi_sq, bart_p_cmp, alpha_opt)
} else {
  ""
}

adequacy_note <- if (adequacy_ok) {
  "Data adequacy is acceptable for EFA."
} else {
  "Note: Data adequacy may be insufficient for reliable EFA."
}

interpretation <- sprintf(
  paste0(
    "EFA with %d factor(s) extracted using %s and %s rotation. ",
    "%s%s",
    "Cumulative variance explained by %d factor(s): %.1f%%. ",
    "%s"
  ),
  nfactors, fm_opt, rotation_opt,
  kmo_str, bart_str,
  nfactors, var_pct,
  adequacy_note
)

# ---- Compose result ----

result <- list(
  adequacy           = adequacy,
  eigenvalues        = eigenvalues,
  variance_explained = variance_explained,
  loadings           = loadings_list,
  communalities      = as.list(communalities),
  uniqueness         = as.list(uniqueness),
  n_factors          = nfactors,
  rotation           = rotation_opt,
  fm                 = fm_opt,
  n                  = n_obs,
  n_vars             = ncol(df),
  variable_names     = colnames(df),
  interpretation     = interpretation
)

result

File: engines/python-scripts/efa.py
"""
Exploratory Factor Analysis (EFA) Script
=========================================
Expects the following variables in the execution namespace (injected by the engine):

  data : list[dict] | dict[str, list] | pd.DataFrame
      The dataset to analyse. Each key is a variable name; values are lists of numbers.

  variables : list[str]
      Names of numeric columns to include. Minimum 3 required.

  nFactors : int | None
      Number of factors to extract. If None or 0, uses the factor_analyzer
      default (factors with eigenvalue > 1). Default: None.

  rotation : str
      Rotation method. One of: "varimax" | "promax" | "oblimin" | "none".
      Default: "varimax".

  extractionMethod : str
      Extraction method. One of: "minres" | "ml" | "principal".
      Default: "minres".

  alpha : float
      Significance level for Bartlett's test. Default: 0.05.

Result structure
----------------
{
  "n": int,
  "n_vars": int,
  "variable_names": list[str],
  "adequacy": {
    "kmo": {
      "overall": float,
      "per_variable": {var: float, ...},
      "interpretation": str
    },
    "bartlett": {
      "chi_square": float,
      "df": int,
      "p_value": float,
      "significant": bool,
      "note": str
    }
  },
  "n_factors": int,
  "rotation": str,
  "extraction_method": str,
  "eigenvalues": list[float],
  "variance_explained": [
    {"factor": "F1", "ss_loadings": float, "prop_var": float, "cumul_var": float},
    ...
  ],
  "loadings": [
    {"variable": str, "F1": float, "F2": float, ...},
    ...
  ],
  "communalities": {var: float, ...},
  "uniqueness": {var: float, ...},
  "interpretation": str
}
"""

from __future__ import annotations

import numpy as np
import pandas as pd
from scipy import stats as scipy_stats


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _to_dataframe(raw) -> pd.DataFrame:
    """Normalise various input shapes to a DataFrame."""
    if isinstance(raw, pd.DataFrame):
        return raw
    if isinstance(raw, list):
        return pd.DataFrame(raw)
    if isinstance(raw, dict):
        return pd.DataFrame(raw)
    raise TypeError(f"Unsupported data type for EFA: {type(raw)}")


def _kmo_interpret(msa: float) -> str:
    if msa >= 0.90:
        return "Marvelous"
    elif msa >= 0.80:
        return "Meritorious"
    elif msa >= 0.70:
        return "Middling"
    elif msa >= 0.60:
        return "Mediocre"
    elif msa >= 0.50:
        return "Miserable"
    else:
        return "Unacceptable"


def _compute_kmo(corr_matrix: np.ndarray, var_names: list[str]) -> dict:
    """
    Compute Kaiser-Meyer-Olkin (KMO) measure of sampling adequacy.
    Uses the partial correlation approach.
    """
    n_vars = corr_matrix.shape[0]
    try:
        corr_inv = np.linalg.inv(corr_matrix)
    except np.linalg.LinAlgError:
        # Singular matrix - regularise with a small diagonal
        corr_inv = np.linalg.pinv(corr_matrix)

    # Partial correlation matrix
    diag_inv = np.sqrt(np.diag(corr_inv))
    partial_corr = np.zeros_like(corr_inv)
    for i in range(n_vars):
        for j in range(n_vars):
            if i != j:
                partial_corr[i, j] = -corr_inv[i, j] / (diag_inv[i] * diag_inv[j])

    # KMO per variable
    corr_sq = corr_matrix ** 2
    partial_sq = partial_corr ** 2

    np.fill_diagonal(corr_sq, 0)
    np.fill_diagonal(partial_sq, 0)

    sum_corr = corr_sq.sum(axis=1)
    sum_partial = partial_sq.sum(axis=1)

    kmo_per = sum_corr / (sum_corr + sum_partial)
    kmo_overall = corr_sq.sum() / (corr_sq.sum() + partial_sq.sum())

    per_var = {var_names[i]: round(float(kmo_per[i]), 4) for i in range(n_vars)}

    return {
        "overall": round(float(kmo_overall), 4),
        "per_variable": per_var,
        "interpretation": _kmo_interpret(float(kmo_overall)),
    }


def _compute_bartlett(corr_matrix: np.ndarray, n_obs: int, alpha: float) -> dict:
    """Bartlett's Test of Sphericity."""
    n_vars = corr_matrix.shape[0]
    det = float(np.linalg.det(corr_matrix))
    det = max(det, 1e-300)  # Guard against log(0)

    chi_sq = -(n_obs - 1 - (2 * n_vars + 5) / 6) * np.log(det)
    df = int(n_vars * (n_vars - 1) / 2)
    p_value = float(scipy_stats.chi2.sf(chi_sq, df))
    significant = p_value < alpha

    return {
        "chi_square": round(float(chi_sq), 4),
        "df": df,
        "p_value": round(p_value, 8),
        "significant": significant,
        "note": (
            "Bartlett's test is significant: factor analysis is appropriate."
            if significant
            else "Bartlett's test is NOT significant: factor analysis may not be appropriate."
        ),
    }


def _run_efa_factor_analyzer(
    df_clean: pd.DataFrame,
    n_factors: int,
    rotation: str,
    method: str,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Run EFA using the factor_analyzer package.
    Returns (loadings, communalities, eigenvalues_from_fa).
    """
    from factor_analyzer import FactorAnalyzer

    rotation_arg = None if rotation == "none" else rotation
    fa = FactorAnalyzer(n_factors=n_factors, rotation=rotation_arg, method=method)
    fa.fit(df_clean.values)

    loadings = fa.loadings_
    communalities = fa.get_communalities()
    ev, _ = fa.get_eigenvalues()
    return loadings, communalities, ev


def _run_efa_sklearn_fallback(
    df_clean: pd.DataFrame,
    n_factors: int,
    rotation: str,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Fallback EFA using principal component extraction (no factor_analyzer).
    Only supports varimax-style rotation via scipy.
    """
    from sklearn.decomposition import PCA
    from scipy.stats import ortho_group

    pca = PCA(n_components=n_factors)
    pca.fit(df_clean.values)
    loadings = pca.components_.T  # shape: (n_vars, n_factors)

    # Scale by sqrt(eigenvalues) to get factor loadings
    eigenvalues_full = pca.explained_variance_
    loadings = loadings * np.sqrt(eigenvalues_full)

    communalities = np.sum(loadings ** 2, axis=1)

    # All eigenvalues from correlation matrix
    corr_matrix = np.corrcoef(df_clean.values.T)
    ev = np.linalg.eigvalsh(corr_matrix)[::-1]

    return loadings, communalities, ev


# ---------------------------------------------------------------------------
# Resolve inputs from injected namespace
# ---------------------------------------------------------------------------

# Get data dictionary (injected by wrapper)
_data_raw = data if "data" in dir() else {}  # noqa: F821

# Get variable names from the 'variables' slot (injected by wrapper)
try:
    _variables = list(variables)  # noqa: F821
except NameError:
    raise ValueError("Variable 'variables' is required - specify column names to analyze")

if len(_variables) == 0:
    raise ValueError("At least one variable must be specified in 'variables'")

# nFactors: 0 or absent means auto-detect (Kaiser criterion)
try:
    _nf_raw = nFactors  # noqa: F821
    _n_factors_req = int(_nf_raw) if _nf_raw is not None else 0
except NameError:
    _n_factors_req = 0

# rotation
try:
    _rotation = str(rotation).lower()  # noqa: F821
except NameError:
    _rotation = "varimax"

# extractionMethod
try:
    _method_raw = str(extractionMethod).lower()  # noqa: F821
except NameError:
    _method_raw = "minres"

# alpha
try:
    _alpha = float(alpha)  # noqa: F821
except NameError:
    _alpha = 0.05

# missingValues
try:
    _missing_values = str(missingValues)  # noqa: F821
except NameError:
    _missing_values = "exclude-listwise"

# Validate rotation
_valid_rotations = {"varimax", "promax", "oblimin", "none", "quartimax"}
if _rotation not in _valid_rotations:
    _rotation = "varimax"

# Map extraction method labels (spec uses "pa"; factor_analyzer uses "principal")
_method_map = {"minres": "minres", "ml": "ml", "principal": "principal", "pa": "principal"}
_method = _method_map.get(_method_raw, "minres")

# ---------------------------------------------------------------------------
# Build analysis dataframe
# ---------------------------------------------------------------------------

df_full = _to_dataframe(_data_raw)

if _variables:
    _missing = [v for v in _variables if v not in df_full.columns]
    if _missing:
        raise ValueError(f"Variables not found in data: {_missing}")
    df_num = df_full[_variables].copy()
else:
    df_num = df_full.select_dtypes(include="number").copy()

# Coerce all columns to numeric
for col in df_num.columns:
    df_num[col] = pd.to_numeric(df_num[col], errors="coerce")

# Handle missing values
if _missing_values == "pairwise":
    # Keep all rows; correlation matrix will use pairwise complete obs
    df_clean = df_num.copy()
    corr_method = "pearson"  # pandas uses pairwise by default for corr()
else:
    # Listwise deletion (default)
    df_clean = df_num.dropna()

var_names = list(df_clean.columns)
n_vars = len(var_names)
n_obs = len(df_clean)

if n_vars < 3:
    raise ValueError(f"EFA requires at least 3 numeric variables; only {n_vars} provided.")

if n_obs < n_vars + 1:
    raise ValueError(
        f"Insufficient observations ({n_obs}) for {n_vars} variables. "
        f"Need at least {n_vars + 1} complete cases."
    )

# ---------------------------------------------------------------------------
# Adequacy tests
# ---------------------------------------------------------------------------

corr_matrix = df_clean.corr().values
kmo_result = _compute_kmo(corr_matrix, var_names)
bartlett_result = _compute_bartlett(corr_matrix, n_obs, _alpha)

adequacy = {
    "kmo": kmo_result,
    "bartlett": bartlett_result,
}

# ---------------------------------------------------------------------------
# Determine number of factors
# ---------------------------------------------------------------------------

# Full eigenvalues from correlation matrix for scree / Kaiser rule
_eig_full = np.linalg.eigvalsh(corr_matrix)[::-1]

if _n_factors_req is None or _n_factors_req <= 0:
    # Kaiser criterion: eigenvalues > 1
    _n_factors = max(1, int(np.sum(_eig_full > 1.0)))
else:
    _n_factors = _n_factors_req

_max_factors = max(1, (n_vars - 1) // 2)
if _n_factors > _max_factors:
    _n_factors = _max_factors

# ---------------------------------------------------------------------------
# Run EFA
# ---------------------------------------------------------------------------

try:
    _loadings, _communalities, _ev_fa = _run_efa_factor_analyzer(
        df_clean, _n_factors, _rotation, _method
    )
    _used_fallback = False
except ImportError:
    _loadings, _communalities, _ev_fa = _run_efa_sklearn_fallback(
        df_clean, _n_factors, _rotation
    )
    _used_fallback = True

_uniqueness = 1.0 - _communalities

# ---------------------------------------------------------------------------
# Build output structures
# ---------------------------------------------------------------------------

_factor_names = [f"F{i + 1}" for i in range(_n_factors)]

# Loadings table
loadings_out = []
for i, var in enumerate(var_names):
    row = {"variable": var}
    for j, fname in enumerate(_factor_names):
        row[fname] = round(float(_loadings[i, j]), 4)
    loadings_out.append(row)

# Communalities and uniqueness
communalities_out = {var_names[i]: round(float(_communalities[i]), 4) for i in range(n_vars)}
uniqueness_out = {var_names[i]: round(float(_uniqueness[i]), 4) for i in range(n_vars)}

# Eigenvalues (from full correlation matrix)
eigenvalues_out = [round(float(v), 4) for v in _eig_full]

# Variance explained
_ss = np.sum(_loadings ** 2, axis=0)
_prop = _ss / n_vars
_cumul = np.cumsum(_prop)
variance_explained_out = [
    {
        "factor": _factor_names[i],
        "ss_loadings": round(float(_ss[i]), 4),
        "prop_var": round(float(_prop[i]), 4),
        "cumul_var": round(float(_cumul[i]), 4),
    }
    for i in range(_n_factors)
]

# ---------------------------------------------------------------------------
# Interpretation
# ---------------------------------------------------------------------------

_adequacy_ok = kmo_result["overall"] >= 0.60 and bartlett_result["significant"]
_var_pct = round(float(_cumul[_n_factors - 1]) * 100, 1)
_fallback_note = " (sklearn PCA fallback; install factor_analyzer for full EFA)" if _used_fallback else ""

interpretation = (
    f"EFA with {_n_factors} factor(s) extracted using {_method}{_fallback_note} "
    f"and {_rotation} rotation. "
    f"KMO = {kmo_result['overall']:.3f} ({kmo_result['interpretation']}). "
    f"Bartlett's test: chi2({bartlett_result['df']}) = {bartlett_result['chi_square']:.2f}, "
    f"p {'<' if bartlett_result['significant'] else '>='} {_alpha:.4f}. "
    f"Cumulative variance explained by {_n_factors} factor(s): {_var_pct:.1f}%. "
    + (
        "Data adequacy is acceptable for EFA."
        if _adequacy_ok
        else "Note: Data adequacy may be insufficient for reliable EFA."
    )
)

# ---------------------------------------------------------------------------
# Compose result
# ---------------------------------------------------------------------------

result = {
    "n": n_obs,
    "n_vars": n_vars,
    "variable_names": var_names,
    "adequacy": adequacy,
    "n_factors": _n_factors,
    "rotation": _rotation,
    "extraction_method": _method,
    "eigenvalues": eigenvalues_out,
    "variance_explained": variance_explained_out,
    "loadings": loadings_out,
    "communalities": communalities_out,
    "uniqueness": uniqueness_out,
    "interpretation": interpretation,
}
